ingester/src/querier_handler.rs

```diff
@@ -1,74 +1,231 @@
+    ))]
+    TableNotFound {
+        sequencer_id: SequencerId,
+        namespace_name: String,
+        table_name: String,
+    },
+
+    #[snafu(display("Error snapshotting non-persisting data: {}", source))]
+    SnapshotNonPersistingData { source: data::Error },
+
+    #[snafu(display("Error concating same-schema record bacthes: {}", source))]
+    ConcatBatches { source: arrow::error::ArrowError },
 }
 
 /// A specialized `Error` for Ingester's Query errors
 pub type Result<T, E = Error> = std::result::Result<T, E>;
 
 /// Return data to send as a response back to the Querier per its request
 pub async fn prepare_data_to_querier(
-    _request: &IngesterQueryRequest,
+    ingest_data: Arc<IngesterData>,
+    request: &IngesterQueryRequest,
 ) -> Result<IngesterQueryResponse> {
-    // Steps:
-    //  1. Read the IngesterData to get TableData for the given request's table
-    //  3. For each PartitionData of the TableData above, get its DataBuffer by:
-    //     3.1. Snapshot the buffer by invoking DataBuffer::snapshot
-    //          This snapshot will not inlcude any tombstones because if there are tomstones
-    //          right before this, it would trigger the sanpshot creation already
-    //     3.2. Converting all snapshots to a QueryableBatch that won't have any tombstones.
-    //     3.3. Get PersistingBatch's QueryableBatches and add tomstones in the list `deletes_during_persisting` if any
-    //     3.4. For each QueryableBatch produced in 3.2 and 3.3, invoke the avaialble `query` function to filter data per request.
-    //          . The request's columns, time range, and predicates will be applied at this step. Each will return a RecordBatch.
-    //     3.5. Put all RecordBatches in a stream
-    //  4. The stream is a part of the returned result
-
-    let schema = Arc::new(ArrowSchema::new(vec![]));
-    let stream = datafusion::physical_plan::EmptyRecordBatchStream::new(Arc::clone(&schema));
+    // ------------------------------------------------
+    //  Read the IngesterData to get TableData for the given request's table
+    let table_data = ingest_data.table_data(
+        request.sequencer_id,
+        request.namespace.as_str(),
+        request.table.as_str(),
+    );
+    if table_data.is_none() {
+        return Err(Error::TableNotFound {
+            sequencer_id: request.sequencer_id,
+            namespace_name: request.namespace.clone(),
+            table_name: request.table.clone(),
+        });
+    }
+
+    // ------------------------------------------------
+    // Accumulate data from each partition
+    let table_data = table_data.unwrap();
+    let partition_keys = table_data.partition_keys();
+
+    // Make Filters
+    let mut selection: Selection<'_> = Selection::All;
+    let mut sel = vec![];
+    if !request.columns.is_empty() {
+        for col in &request.columns {
+            sel.push(col.as_str());
+        }
+        selection = Selection::Some(&sel);
+    }
```

Comment:
This is potentially doing multiple allocations of Vec as it grows; better would be using a `collect` to allocate the exact size once. This is a little tricky with `Selection::Some` needing a slice, but I think this is a little better at least:

```suggestion
    let selection_columns: Vec<_> = request.columns.iter().map(String::as_str).collect();
    let selection = if selection_columns.is_empty() {
        Selection::All
    } else {
        Selection::Some(&selection_columns)
    };
```

Reply:
Yeah, I got compile error when trying to create a slice :)